### PR TITLE
[RLlib] Cleanup examples folder (vol 23): Float16 training support and new example script.

### PR DESCRIPTION
**Pull request by sven1977**
Cleanup examples folder (vol 23): Float16 training support and new example script.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/package_ref/learner.rst
Comment (sven1977) [inline doc/source/rllib/package_ref/learner.rst:94]: I changed this API name to be even more clear. This method computes one(!) loss per RLModule (in a MultiRLModule) inside the Learner.

Got rid of the confusing `TOTAL_LOSS` key. We compute this now in the default implementation of `compute_gradients`.

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py
Comment (sven1977) [inline rllib/algorithms/ppo/torch/ppo_torch_learner.py:102]: simplify

File: rllib/core/learner/learner.py
Comment (sven1977) [inline rllib/core/learner/learner.py:1628]: Emulate the old behavior, in case users use this in their tests or other code.

File: rllib/BUILD
Comment (simonsays1980) [inline rllib/BUILD:2979]: Interesting. Is this even used by users?

File: rllib/algorithms/algorithm_config.py
Comment (simonsays1980) [inline rllib/algorithms/algorithm_config.py:3202]: Small nit: Remove "the" at the end.

File: rllib/algorithms/algorithm_config.py
Comment (simonsays1980) [inline rllib/algorithms/algorithm_config.py:3207]: The description is not clear enough imo. For what is it used?

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py
Comment (simonsays1980) [inline rllib/algorithms/ppo/torch/ppo_torch_learner.py:54]: Could we potentially also mask with this `mask` values in the observations that are not available in a certain step of the environment, e.g. different number of entities at different steps of a game?

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py
Comment (simonsays1980) [inline rllib/algorithms/ppo/torch/ppo_torch_learner.py:103]: Niceee!

File: rllib/core/learner/learner.py
Comment (simonsays1980) [inline rllib/core/learner/learner.py:219]: Remove `_hps`. Hyperparameters are not used anymore.

File: rllib/core/learner/learner.py
Comment (simonsays1980) [inline rllib/core/learner/learner.py:857]: Very important hint! Awesome!

File: rllib/core/learner/learner.py
Comment (simonsays1980) [inline rllib/core/learner/learner.py:856]: Let's either refer to the docs about this or to some examples. Just from the text this does not become very clear.

File: rllib/core/learner/learner.py
Comment (simonsays1980) [inline rllib/core/learner/learner.py:899]: Can `compute_losses` call internally `compute_loss_for_module` or is the latter called nevertheless and losses would be computed two times?

File: rllib/core/learner/tf/tf_learner.py
Comment (simonsays1980) [inline rllib/core/learner/tf/tf_learner.py:303]: Do we even want this here and motivate users to write TF algorithms?

File: rllib/core/learner/torch/torch_learner.py
Comment (simonsays1980) [inline rllib/core/learner/torch/torch_learner.py:122]: Don't we want to pass in `kwargs` to `Adam`?

File: rllib/core/learner/torch/torch_learner.py
Comment (simonsays1980) [inline rllib/core/learner/torch/torch_learner.py:162]: Awesome!!!

File: rllib/examples/gpus/float16_training_and_inference.py
Comment (simonsays1980) [inline rllib/examples/gpus/float16_training_and_inference.py:4]: Absolutely awesome!!! This is such a great example!

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome PR. It would be nice to add some more documentation of why certain components are needed and how the new loss setup works now.

File: rllib/BUILD
Comment (sven1977) [reply rllib/BUILD:2979]: Yes. Actually a customer asked for it ;)
It does get interesting for large models (see also many efforts training LLMs with super compressed precisions down to bfloat16) and multi-agent. I think if one can stabilize this, it's very useful. Another example script with mixed-precision training (and float16 inference on the EnvRunners) is in the making ...

File: rllib/algorithms/algorithm_config.py
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:3202]: done

File: rllib/algorithms/algorithm_config.py
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:3207]: enhanced and added a link to the torch docs

File: rllib/algorithms/ppo/torch/ppo_torch_learner.py
Comment (sven1977) [reply rllib/algorithms/ppo/torch/ppo_torch_learner.py:54]: That should work. But you would have to manually change that in the Learner connector (where this column is being produced and added to the batch).

File: rllib/core/learner/learner.py
Comment (sven1977) [reply rllib/core/learner/learner.py:219]: Great catch! Will fix.

Comment (sven1977) [reply rllib/core/learner/learner.py:219]: fixed

File: rllib/core/learner/learner.py
Comment (sven1977) [reply rllib/core/learner/learner.py:856]: I can link to the custom loss example ...

Comment (sven1977) [reply rllib/core/learner/learner.py:856]: done

File: rllib/core/learner/tf/tf_learner.py
Comment (sven1977) [reply rllib/core/learner/tf/tf_learner.py:303]: Good point! But .... DreamerV3 :|

File: rllib/core/learner/learner.py
Comment (sven1977) [reply rllib/core/learner/learner.py:899]: The default implementation of `compute_losses` calls n times `compute_loss_for_module`, where n is the number of RLModules within the Learner's MultiRLModule. So nothing is computed twice.

If you have a complex multi-agent case, you should override `compute_losses`, in which case the n calls to `compute_loss_for_module` will NOT be made.

File: rllib/core/learner/torch/torch_learner.py
Comment (sven1977) [reply rllib/core/learner/torch/torch_learner.py:122]: Used to think the same way.

BUT that would make configuring custom optimizers again very non-pythonic and very yaml'ish. The user would then have to provide a class/type and some kwargs, but then has no chance to customize anything else within the optimizer setup process. Think of a user having to configure two optimizers, or three. Where do you make these options available in the config, then? What if the user needs different optimizers per module?

File: rllib/examples/gpus/float16_training_and_inference.py
Comment (sven1977) [reply rllib/examples/gpus/float16_training_and_inference.py:4]: Thanks a ton!! Yes, and it actually works :D (only with adding all these weird tricks to stabilize learning). But I agree, it's super nice to see how users can customize not just one component at the same time, but several and then achieve a very high degree of customizability w/o us having to add feature after feature to the lib.

**Comment by can-anyscale** [comment]
Broke https://github.com/ray-project/ray/issues/45088 and is blocking the release, i'm reverting to unblock

**Pull request merged**
